Training/Program.cs

```diff
@@ -1,33 +1,120 @@
-// Program on main branch.
+// Program.cs
+// Program to implement a file name parser using a Mealy State Machine.
 // ------------------------------------------------------------------------------------------------
-namespace Training {
-   #region class Program --------------------------------------------------------------------------
-   internal class Program {
-      #region Constructors ------------------------------------------
-      #endregion
-
-      #region Method ------------------------------------------------
-      static void Main (string[] args) {
-         Console.WriteLine ("Hello, World!");
-      }
-      #endregion
+using static System.Console;
+using static Training.EState;
+namespace Training;
+
+class Program {
+   static void Main (string[] args) {
+      Dictionary<string, (string, string, string, string, string)> validPaths = new () {
+         { "C:\\WorkGIT\\Dakshita\\Training.sln", ("c", "workgit", "dakshita", "training.sln", ".sln")},
+         { "C:/words/words.txt", ("c", "words", "", "words.txt", ".txt")}
+      };
+      string[] invalidPaths = { "suppppp", "Ca:/workgit/hjh.txt", "c:\\Dakshita123.md", "C:\\workgit\\dakshita", "c:/dakshita.txt" };
+      foreach (var kvp in validPaths)
+         Validate (kvp.Key);
+      foreach (string path in invalidPaths)
+         Validate (path);
 
-      #region Operators ---------------------------------------------
-      #endregion
+      static void Validate (string path) =>
+         WriteLine (Path.PathParse (path, out (string, string, string, string, string) ActOutput) + "\t" + ActOutput);
+
+      //Path fp = new ("C:\\WorkGIT\\Dakshita\\Training.sln");
+      //WriteLine (fp);
+      //WriteLine (fp.FilePath);
+   }
+}
 
-      #region Implementation ----------------------------------------
-      #endregion
+#region class Path --------------------------------------------------------------------------
+public class Path {
+   #region Constructors ------------------------------------------
+   /// <summary>Creates new instance of a filepath</summary>
+   public Path (string path) => mPath = path.Trim ().ToLower ();
+   #endregion
 
-      #region Nested Types ------------------------------------------
-      #endregion
+   #region Properties ------------------------------------------
+   /// <summary>If path is valid, returns components of a file's path, else returns empty strings.</summary>
+   public (string Drive, string Dir, string Path, string Filename, string Ext) FilePath {
+      get {
+         // FilePath is computed only once for 1 instance of class (lazy evaluation).
+         if (mComputed)
+            return mFilePath;
+         else {
+            mValid = PathParse (mPath, out (string drive, string dir, string path, string filename, string ext) f);
+            mFilePath = mValid ? f : ("", "", "", "", "");
+            mComputed = true;
+            nComputations++;
+            return mFilePath;
+         }

```

Comment:
Better way is as follows:
```suggestion
        if (mFilePath.drive == null) { 
           mValid = PathParse (mPath, out (string drive, string dir, string path, string filename, string ext) f);
            mFilePath = mValid ? f : ("", "", "", "", "");
            mComputed = true;
            nComputations++;
        }       
            return mFilePath;

```
Above code avoids two return mFilePath, mComputed variable as drive should not be null even if is not valid.

Reply:
But what if the input start with something like '@' or some other non-alphabet?